### ISSUE TITLE
No simple way to create a windows::core::Error from HRESULT/NTSTATUS/WIN32_ERROR constants

### ISSUE THREAD

**Issue by roblabla** [microsoft/windows-rs#1578]
Currently, the only way I could find to turn `HRESULT` directly into an `Error` without going through a `Result`. For instance, this is the only way I found to turn an E_OUTOFMEMORY into an error:

```rust
let err = E_OUTOFMEMORY
    .ok()
    .map(|_| unreachable!())
    .unwrap_err();
```

This is kinda ugly though, as it forces a lot of "panicky" unwrapping code. What's worse, this can't be done from a `const fn`, because `HRESULT::ok` is a fairly complex function that cannot be made const trivially. This is somewhat annoying, as it prevents writing some safe wrappers around win32 functions that fetch the allocation size, allocate, and try the function in a loop, such as this:

```rust
pub fn get_extended_udp_table_owner_pid_ipv4(order: bool) -> hl_windows::core::Result<MibUdpTableOwnerPid> {
    let class = UDP_TABLE_OWNER_PID;

    // Default to allocating enough for 32 connections.
    let mut requested_size = (32 * size_of::<MIB_UDPROW_OWNER_PID>() + size_of::<MIB_UDPTABLE_OWNER_PID>()) as u32;

    // Try to fetch size a couple times before failing.
    for _ in 0..10 {
        let udptable = MibUdpTableOwnerPid::alloc(requested_size as usize)?;

        let res = unsafe {
            // Safety: udptable.inner is an allocated ptr by construction of MibUdpTableOwnerPid.
            // It is also of the right type (AF_INET+UDP_TABLE_OWNER_PID -> MIB_UDP_TABLE_OWNER_PID)
            GetExtendedUdpTable(
                udptable.inner as _,
                &mut requested_size,
                order,
                AF_INET,
                class,
                0,
            )
        };
        let win32_err = WIN32_ERROR(res);
        let res = HRESULT::from(win32_err).ok();

        match res {
            Ok(()) => return Ok(udptable),
            Err(_err) if win32_err == ERROR_INSUFFICIENT_BUFFER => (),
            Err(err) => return Err(err),
        }
    }

    ERROR_INSUFFICIENT_BUFFER.ok().map(|_| unreachable!())
}
```

Of course, I could restructure my code to avoid having to create the result myself (e.g. by having a `loop` and conting myself), but I find this structure much easier to read and reason about.

Having a way to turn an `HRESULT` directly into an `Error` without going through `Result` would make the code a tiny bit better.

**Comment by kennykerr** [comment]
Not exactly the same thing but #1483 includes switching the default conversion to be fast by default (no WinRT error reporting) and should work with a simple "into".

**Comment by kennykerr** [comment]
I have addressed this with #1585 - let me know if that looks good to you.
